docs/notebooks/demo/demo-utilities-lib/dgan_tuner_utils.py

```diff
@@ -85,8 +137,85 @@ def undo_padding(group, event_column, pad_value="[END]"):
     return unpadded_group
 
 
+def compute_actions_from_events(df, example_id_column, event_column):
+    """
+    Computes the event for each row in the DataFrame based on the changes
+    in the event_column values for each unique identifier in example_id_column. The action
+    is determined as 'start', 'increment', or 'unchanged'.
+    
+    Args:
+    - df (pd.DataFrame): The input DataFrame to process.
+    - example_id_column (str): The column name acting as the unique session identifier.
+    - event_column (str): The column name representing the event count.
+    
+    Returns:
+    - pd.DataFrame: The DataFrame with an updated 'event_column' column.
+    """
+    # Sort df by example_id_column and event_column to ensure the sequence is correct
+    df_sorted = df.sort_values(by=[example_id_column, event_column]).reset_index(drop=True)
+
+    df_sorted['tmp'] = None
+
+    previous_id = None
+    previous_count = None
+
+    for index, row in df_sorted.iterrows():
+        if row[example_id_column] != previous_id:  # New session
+            df_sorted.at[index, 'tmp'] = 'start'
+        else:  # Same session as previous row
+            if row[event_column] == previous_count:  # No change in event count
+                df_sorted.at[index, 'tmp'] = 'unchanged'
+            elif row[event_column] > previous_count:  # Event count has increased
+                df_sorted.at[index, 'tmp'] = 'increment'
+            # Assuming event_column only increases or stays the same,
+            # so no else case for decrement
+
+        # Update previous values for the next iteration
+        previous_id = row[example_id_column]
+        previous_count = row[event_column]
+    
+    df_sorted[event_column] =  df_sorted['tmp']
+    df_sorted.drop('tmp', axis=1, inplace=True)
+    
+    return df_sorted.reset_index(drop=True)
+
+
+def compute_events_from_actions(df, event_column, pad_value="[END]"):
+    """
+    Processes the given DataFrame by removing rows with a specified action in event_column set to 'pad_value',
+    initializing a new 'new_funnel_count' column, and updating it based on event_column values.
+    
+    Args:
+    - df (pd.DataFrame): The input DataFrame to process.
+    - event_column (str): The name of the column containing event actions.
+    
+    Returns:
+    - pd.DataFrame: The processed DataFrame with updated event_column column.
+    """
+        
+    df_new = df[df[event_column] != pad_value].reset_index(drop=True).copy()
+
+    # Initialize the new_funnel_count column
+    df_new['tmp'] = 0
+
+    # Apply actions to calculate new_funnel_count
+    for i, row in df_new.iterrows():
+        if row[event_column] == 'start':
```

Comment:
question: This logic assumes every sequences begins with `start`, and all other `event_column` values are `increment` or `unchanged`. Is there some other place we're tossing out synthetic sequences where that's not true? Or does dgan get this pattern correct (nearly) all the time?

Reply:
Yes, (surprisingly?) DGAN gets that pattern correct all the time. Each sequence start with a funnel_count = 'start'